include/small/models/Darknet.hpp

```diff
@@ -0,0 +1,530 @@
+        ActivationType activation = ActivationType::NONE; 
+
+        int last_pos = cfg_file.tellg();
+        std::string line;
+        while(getline(cfg_file, line)) {
+
+            // skip empty lines and comments
+            if(line.empty() || line.at(0) == '#') { last_pos = cfg_file.tellg(); continue; }
+            // stop parsing when we reach the next layer
+            if(line.at(0) == '[') { cfg_file.seekg(last_pos); break; }
+            last_pos = cfg_file.tellg();
+            std::string key = line.substr(0, line.find("="));
+            std::string value = line.substr(line.find("=") + 1);
+            
+
+            /// @todo: shortcut also specifies an activation but its always linear
+            /// ignore activation for now, but will need to support in the future
+            if(key == "from") { from = stoi(value); }
+            else if(key == "activation") { activation = parse_activation(value); (void)activation; }
+            else { std::cerr << "WARNING: unknown key in shortcut layer: " << key << std::endl; }
+        }
+
+        // create shortcut layer
+        size_t total_layers_so_far = this->m_layers.size();
+        AddLayer<BufferT> *shortcut = new AddLayer<BufferT> (input, this->m_layers[total_layers_so_far + from]->output_shape());
+        return shortcut;
+    }
+
+    Layer<BufferT>* parse_route(std::ifstream &cfg_file) {
+
+        #ifdef PARSER_DEBUG_VERBOSE
+        std::cout << "Parsing Route Layer\n";
+        #endif
+
+        std::vector<int> route_layers;
+
+        int last_pos = cfg_file.tellg(); 
+        std::string line;
+        while(getline(cfg_file, line)) {
+
+            // skip empty lines and comments
+            if(line.empty() || line.at(0) == '#') { last_pos = cfg_file.tellg(); continue; }
+            // stop parsing when we reach the next layer
+            if(line.at(0) == '[') { cfg_file.seekg(last_pos); break; }
+            last_pos = cfg_file.tellg();
+            std::string key = line.substr(0, line.find("="));
+            std::string value = line.substr(line.find("=") + 1);
+
+            if(key == "layers") { route_layers = extract_int_array<int>(value); }
+            else { std::cerr << "WARNING: unknown key in route layer: " << key << std::endl; }
+        }
+
+        RouteLayer<BufferT> *route;
+        std::vector<shape_type> inputs;
+
+        for(auto layer : route_layers) {
+            // handle negative indexing
+            if(layer < 0) { inputs.push_back(this->m_layers[this->m_layers.size() + layer]->output_shape()); }
+            else { inputs.push_back(this->m_layers[layer]->output_shape()); }
+        }
+
+        if(inputs.size() == 1) { route = new RouteLayer<BufferT> (inputs[0]); }
+        else if(inputs.size() == 2) { route = new RouteLayer<BufferT> (inputs[0], inputs[1]); }
+        else {
+            std::cerr << "ERROR: route layer must have 1 or 2 layers" << std::endl;
+            throw std::exception(); /// @todo: throw some exception
```

Comment:
consider throwing std::format_error or std::invalid_argument

Reply:
Fixed in [f419425](https://github.com/CMU-SPEED/SMaLLFramework/pull/7/commits/f4194253ec542140623dcae1d5440638c1b3ee99)